### PR TITLE
feature(zp): add dreamerv3 algorithm

### PR DESCRIPTION
**Pull request by zhangpaipai**
## Check List

- [ ] merge the latest version source branch/repo, and resolve all the conflicts
- [ ] pass style check
- [ ] pass all the tests

### CODE REVIEW

File: ding/world_model/utils.py [hunk 1]
```diff
@@ -23,3 +28,255 @@ def get_rollout_length_scheduler(cfg: EasyDict) -> Callable[[int], int]:
...
+        self._threshold = threshold
+
+    def log_prob(self, event):
+        return -(torch.sqrt((event - self.mean) ** 2 + self._threshold ** 2) - self._threshold)
+
+    def mode(self):
+        return self.mean
+
+
+def weight_init(m):
```
Comment (PaParaZz1) [inline ding/world_model/utils.py:239]: add usage comments about this init function

File: ding/world_model/utils.py [hunk 2]
```diff
@@ -23,3 +28,255 @@ def get_rollout_length_scheduler(cfg: EasyDict) -> Callable[[int], int]:
...
+        if self._agg == 'mean':
+            loss = distance.mean(self._dim_to_reduce)
+        elif self._agg == 'sum':
+            loss = distance.sum(self._dim_to_reduce)
+        else:
+            raise NotImplementedError(self._agg)
+        return -loss
+
+
+class ContDist:
```
Comment (PaParaZz1) [inline ding/world_model/utils.py:175]: polish all the class name of distributions we implemented, unify their name format

File: ding/world_model/utils.py [hunk 3]
```diff
@@ -23,3 +28,255 @@ def get_rollout_length_scheduler(cfg: EasyDict) -> Callable[[int], int]:
...
+            distance = (self._mode - symlog(value)) ** 2.0
+            distance = torch.where(distance < self._tol, 0, distance)
+        elif self._dist == 'abs':
+            distance = torch.abs(self._mode - symlog(value))
+            distance = torch.where(distance < self._tol, 0, distance)
+        else:
+            raise NotImplementedError(self._dist)
+        if self._agg == 'mean':
+            loss = distance.mean(self._dim_to_reduce)
+        elif self._agg == 'sum':
```
Comment (PaParaZz1) [inline ding/world_model/utils.py:168]: maybe we can polish the related variable name and make them more readable, such as `agg` to `aggregation`

File: ding/world_model/utils.py [hunk 4]
```diff
@@ -23,3 +28,255 @@ def get_rollout_length_scheduler(cfg: EasyDict) -> Callable[[int], int]:
...
+        return (target * log_pred).sum(-1)
+
+    def log_prob_target(self, target):
+        log_pred = super().logits - torch.logsumexp(super().logits, -1, keepdim=True)
+        return (target * log_pred).sum(-1)
+
+
+class SymlogDist():
+
+    def __init__(self, mode, dist='mse', agg='sum', tol=1e-8, dim_to_reduce=[-1, -2, -3]):
```
Comment (PaParaZz1) [inline ding/world_model/utils.py:143]: add python typing lint

File: ding/world_model/utils.py [hunk 5]
```diff
@@ -23,3 +28,255 @@ def get_rollout_length_scheduler(cfg: EasyDict) -> Callable[[int], int]:
         return lambda x: cfg.rollout_length
     else:
         raise KeyError("not implemented key: {}".format(cfg.type))
+
+
+def symlog(x):
```
Comment (PaParaZz1) [inline ding/world_model/utils.py:33]: move this two functions to `ding/rl_utils/value_rescale.py`

File: ding/world_model/utils.py [hunk 6]
```diff
@@ -23,3 +28,255 @@ def get_rollout_length_scheduler(cfg: EasyDict) -> Callable[[int], int]:
...
+    return torch.sign(x) * (torch.exp(torch.abs(x)) - 1.0)
+
+
+class SampleDist:
+
+    def __init__(self, dist, samples=100):
+        self._dist = dist
+        self._samples = samples
+
+    @property
```
Comment (PaParaZz1) [inline ding/world_model/utils.py:47]: why add this method

File: ding/world_model/utils.py [hunk 7]
```diff
@@ -23,3 +28,255 @@ def get_rollout_length_scheduler(cfg: EasyDict) -> Callable[[int], int]:
...
+    def name(self):
+        return 'SampleDist'
+
+    def __getattr__(self, name):
```
Comment (PaParaZz1) [inline ding/world_model/utils.py:51]: why overwrite this method here

File: ding/world_model/utils.py [hunk 8]
```diff
@@ -23,3 +28,255 @@ def get_rollout_length_scheduler(cfg: EasyDict) -> Callable[[int], int]:
...
+class TwoHotDistSymlog():
+
+    def __init__(self, logits=None, low=-20.0, high=20.0, device='cpu'):
+        self.logits = logits
+        self.probs = torch.softmax(logits, -1)
+        self.buckets = torch.linspace(low, high, steps=255).to(device)
+        self.width = (self.buckets[-1] - self.buckets[0]) / 255
+
+    def mean(self):
+        print("mean called")
```
Comment (PaParaZz1) [inline ding/world_model/utils.py:104]: remove debug print

File: ding/world_model/utils.py [hunk 9]
```diff
@@ -23,3 +28,255 @@ def get_rollout_length_scheduler(cfg: EasyDict) -> Callable[[int], int]:
...
+
+    def __getattr__(self, name):
+        return getattr(self._dist, name)
+
+    def entropy(self):
+        return self._dist.entropy()
+
+    def mode(self):
+        _mode = torch.round(self._dist.mean)
+        return _mode.detach() + self._dist.mean - self._dist.mean.detach()
```
Comment (PaParaZz1) [inline ding/world_model/utils.py:213]: why use detach here

File: ding/world_model/model/networks.py [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+        x = features
+        out = self.mlp(x)  # (batch, time, _units=512)
+        mean = self.mean_layer(out)  # (batch, time, 255)
+        if self._std == "learned":
+            std = self.std_layer(out)
+        else:
+            std = self._std
+        if self._dist == "normal":
+            return ContDist(torchd.independent.Independent(torchd.normal.Normal(mean, std), len(self._shape)))
+        if self._dist == "huber":
```
Comment (PaParaZz1) [inline ding/world_model/model/networks.py:66]: elif here

File: ding/model/common/tests/test_encoder.py [hunk 11]
```diff
@@ -47,3 +61,7 @@ def test_impalaconv_encoder(self):
         outputs = model(inputs)
         self.output_check(model, outputs)
         assert outputs.shape == (B, 256)
+
+
+a = TestEncoder()
+a.test_dreamer_conv_encoder()
```
Comment (PaParaZz1) [inline ding/model/common/tests/test_encoder.py:67]: why add these two lines

File: ding/model/common/encoder.py [hunk 12]
```diff
@@ -306,3 +326,41 @@ def forward(self, x):
...
+            self.bias,
+            self.stride,
+            self.padding,
+            self.dilation,
+            self.groups,
+        )
+        return ret
+
+
+class DreamerLayerNorm(nn.Module):
```
Comment (PaParaZz1) [inline ding/model/common/encoder.py:356]: move these two classes to `ding/torch_utils/network`

File: ding/model/common/encoder.py [hunk 13]
```diff
@@ -63,17 +65,35 @@ def __init__(
...
+                layers.append(DreamerLayerNorm(hidden_size_list[i]))
+                layers.append(self.act)
+            else:
+                layers.append(nn.Conv2d(input_size, hidden_size_list[i], kernel_size[i], stride[i], padding[i]))
+                layers.append(self.act)
             input_size = hidden_size_list[i]
-        assert len(set(hidden_size_list[3:-1])) <= 1, "Please indicate the same hidden size for res block parts"
-        for i in range(3, len(self.hidden_size_list) - 1):
-            layers.append(ResBlock(self.hidden_size_list[i], activation=self.act, norm_type=norm_type))
+        if len(self.hidden_size_list) >= len(kernel_size) + 2:
```
Comment (PaParaZz1) [inline ding/model/common/encoder.py:84]: format codes with with a maximum of 120 characters per line

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/model/common/encoder.py [hunk 14]
```diff
@@ -63,17 +65,35 @@ def __init__(
...
+            set(hidden_size_list[len(kernel_size):-1])
+        ) <= 1, "Please indicate the same hidden size for res block parts"
+        for i in range(len(kernel_size), len(self.hidden_size_list) - 1):
+            layers.append(ResBlock(self.hidden_size_list[i - 1], activation=self.act, norm_type=norm_type))
         layers.append(Flatten())
         self.main = nn.Sequential(*layers)
 
         flatten_size = self._get_flatten_size()
-        self.output_size = hidden_size_list[-1]
+        self.output_size = hidden_size_list[-1]  # outside to use
```
Comment (zjowowen) [inline ding/model/common/encoder.py:96]: "# outside to use" remove comment

File: ding/model/common/encoder.py [hunk 15]
```diff
@@ -306,3 +326,41 @@ def forward(self, x):
         if self.final_relu:
             x = torch.relu(x)
         return x
+
+
+class Conv2dSame(torch.nn.Conv2d):
+
+    def calc_same_pad(self, i, k, s, d):
+        return max((math.ceil(i / s) - 1) * s + (k - 1) * d + 1 - i, 0)
```
Comment (zjowowen) [inline ding/model/common/encoder.py:334]: Add data type for i,k,s,d
Add function notation

File: ding/model/common/encoder.py [hunk 16]
```diff
@@ -306,3 +326,41 @@ def forward(self, x):
...
+            self.dilation,
+            self.groups,
+        )
+        return ret
+
+
+class DreamerLayerNorm(nn.Module):
+
+    def __init__(self, ch, eps=1e-03):
+        super(DreamerLayerNorm, self).__init__()
```
Comment (zjowowen) [inline ding/model/common/encoder.py:359]: Add data type

File: ding/world_model/model/networks.py [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+
+from ding.world_model.utils import weight_init, uniform_weight_init, ContDist, Bernoulli, TwoHotDistSymlog, UnnormalizedHuber
+from ding.torch_utils import MLP, fc_block
+
+
+class DenseHead(nn.Module):
+
+    def __init__(
+        self,
+        inp_dim,  # config.dyn_stoch * config.dyn_discrete + config.dyn_deter
```
Comment (zjowowen) [inline ding/world_model/model/networks.py:17]: remove " # config.dyn_stoch * config.dyn_discrete + config.dyn_deter"?

File: ding/world_model/model/networks.py [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+import torch
+from torch import nn
+import torch.nn.functional as F
+from torch import distributions as torchd
+
+from ding.world_model.utils import weight_init, uniform_weight_init, ContDist, Bernoulli, TwoHotDistSymlog, UnnormalizedHuber
+from ding.torch_utils import MLP, fc_block
+
+
+class DenseHead(nn.Module):
```
Comment (zjowowen) [inline ding/world_model/model/networks.py:13]: Tell some information about this class.

File: ding/world_model/model/networks.py [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+        self.mlp.apply(weight_init)
+
+        self.mean_layer = nn.Linear(self._units, np.prod(self._shape))
+        self.mean_layer.apply(uniform_weight_init(outscale))
+
+        if self._std == "learned":
+            self.std_layer = nn.Linear(self._units, np.prod(self._shape))
+            self.std_layer.apply(uniform_weight_init(outscale))
+
+    def forward(self, features, dtype=None):
```
Comment (zjowowen) [inline ding/world_model/model/networks.py:56]: what's the use of dtype?
I can not find where it is used?

Comment (zjowowen) [reply ding/world_model/model/networks.py:56]: I think model module should take one input for forward function.

File: ding/world_model/model/networks.py [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+        x = features
+        out = self.mlp(x)  # (batch, time, _units=512)
+        mean = self.mean_layer(out)  # (batch, time, 255)
+        if self._std == "learned":
+            std = self.std_layer(out)
```
Comment (zjowowen) [inline ding/world_model/model/networks.py:61]: please make sure the std_layer output is always non-negative.
I am afraid there would be dangerous by using only nn.Linear

File: ding/world_model/model/networks.py [hunk 21]
```diff
@@ -0,0 +1,72 @@
+import math
+import numpy as np
+
+import torch
+from torch import nn
+import torch.nn.functional as F
+from torch import distributions as torchd
```
Comment (zjowowen) [inline ding/world_model/model/networks.py:7]: please do not use torchd as short module name. It's not a standard coding format.
It's better to from torch.distributions import XXXXX.
Import only what you need.

File: ding/world_model/utils.py [hunk 22]
```diff
@@ -23,3 +28,255 @@ def get_rollout_length_scheduler(cfg: EasyDict) -> Callable[[int], int]:
...
+        logprob = self._dist.log_prob(sample)
+        return sample[torch.argmax(logprob)][0]
+
+    def entropy(self):
+        sample = self._dist.sample(self._samples)
+        logprob = self.log_prob(sample)
+        return -torch.mean(logprob, 0)
+
+
+class OneHotDist(torchd.one_hot_categorical.OneHotCategorical):
```
Comment (zjowowen) [inline ding/world_model/utils.py:69]: Add class notation

File: ding/world_model/utils.py [hunk 23]
```diff
@@ -23,3 +28,255 @@ def get_rollout_length_scheduler(cfg: EasyDict) -> Callable[[int], int]:
...
+
+def symlog(x):
+    return torch.sign(x) * torch.log(torch.abs(x) + 1.0)
+
+
+def symexp(x):
+    return torch.sign(x) * (torch.exp(torch.abs(x)) - 1.0)
+
+
+class SampleDist:
```
Comment (zjowowen) [inline ding/world_model/utils.py:41]: Add class notation

Comment (zjowowen) [reply ding/world_model/utils.py:41]: Why this class is needed? It seems do not contain any useful information?
You can just rewrite these methods into single line code.

File: ding/world_model/utils.py [hunk 24]
```diff
@@ -23,3 +28,255 @@ def get_rollout_length_scheduler(cfg: EasyDict) -> Callable[[int], int]:
...
+            raise ValueError('need to check')
+        sample = super().sample(sample_shape)
+        probs = super().probs
+        while len(probs.shape) < len(sample.shape):
+            probs = probs[None]
+        sample += probs - probs.detach()
+        return sample
+
+
+class TwoHotDistSymlog():
```
Comment (zjowowen) [inline ding/world_model/utils.py:95]: Add class and function notations

File: ding/world_model/utils.py [hunk 25]
```diff
@@ -23,3 +28,255 @@ def get_rollout_length_scheduler(cfg: EasyDict) -> Callable[[int], int]:
...
+
+    def __init__(self, logits=None, low=-20.0, high=20.0, device='cpu'):
```
Comment (zjowowen) [inline ding/world_model/utils.py:97]: device is set by default. It should not  be explicitly set by some argument.

File: ding/world_model/utils.py [hunk 26]
```diff
@@ -23,3 +28,255 @@ def get_rollout_length_scheduler(cfg: EasyDict) -> Callable[[int], int]:
...
+        return self._dist.mean
+
+    def sample(self, sample_shape=()):
+        return self._dist.rsample(sample_shape)
+
+    def log_prob(self, x):
+        return self._dist.log_prob(x)
+
+
+class Bernoulli:
```
Comment (zjowowen) [inline ding/world_model/utils.py:198]: Add class notation

File: ding/world_model/utils.py [hunk 27]
```diff
@@ -23,3 +28,255 @@ def get_rollout_length_scheduler(cfg: EasyDict) -> Callable[[int], int]:
...
+
+    def log_prob(self, x):
+        _logits = self._dist.base_dist.logits
+        log_probs0 = -F.softplus(_logits)
+        log_probs1 = -F.softplus(-_logits)
+
+        return log_probs0 * (1 - x) + log_probs1 * x
+
+
+class UnnormalizedHuber(torchd.normal.Normal):
```
Comment (zjowowen) [inline ding/world_model/utils.py:226]: Add notation

**Review by PaParaZz1 — APPROVED** [review]
